godot_project/editor/rendering/atom_autopose_preview/atom_autopose_preview.gd

```diff
@@ -0,0 +1,161 @@
+var _hovered_candidate: AtomCandidate
+var _camera: Camera3D
+var _camera_last_transform: Transform3D
+
+
+func set_atomic_number(in_atomic_number: int) -> void:
+	_new_atomic_number = in_atomic_number
+	queue_redraw()
+
+
+func set_bond_order(in_bond_order: int) -> void:
+	_new_bond_order = in_bond_order
+	queue_redraw()
+
+
+func set_candidates(in_candidates: Array[AtomCandidate]) -> void:
+	_candidates = in_candidates
+	queue_redraw()
+
+
+func get_hovered_candidate_or_null() -> AtomCandidate:
+	return _hovered_candidate
+
+
+func _input(event: InputEvent) -> void:
+	if visible and event is InputEventMouse:
+		var hovered: AtomCandidate = null
+		var hovered_distance_sqrd: float = INF
+		const MIN_DISTANCE_SQRD: float = 15*15
+		for candidate: AtomCandidate in _candidates:
+			var distance_sqrd: float = event.position.distance_squared_to(candidate.pos_2d_cache)
+			if distance_sqrd < MIN_DISTANCE_SQRD:
+				if distance_sqrd < hovered_distance_sqrd:
+					hovered_distance_sqrd = distance_sqrd
+					hovered = candidate
+		if hovered != _hovered_candidate:
+			_hovered_candidate = hovered
+			queue_redraw()
+
+
+func _ready() -> void:
+	_camera = get_viewport().get_camera_3d()
+
+
+func _process(delta: float) -> void:
+	if visible and is_instance_valid(_camera) and _camera.global_transform != _camera_last_transform:
+		_camera_last_transform = _camera.global_transform
+		queue_redraw()
+
+
+# Visual language:
+# Line: bond in near connected atom plane
+# Wedges (slim triagles): atoms above the connected atom plane
+# Dashes: atoms below the connected atom plane
+func _draw() -> void:
+	# If control is being drawn we assume candidates belongs to the current workspace context
+	var font: Font = get_theme_font(&"default_font")
+	_camera = get_viewport().get_camera_3d()
+	var camera_up_vector: Vector3 = _camera.basis.y
+	var curr_atom_data: ElementData = PeriodicTable.get_by_atomic_number(_new_atomic_number)
+	const CARBON: int = 6
+	var carbon_data: ElementData = PeriodicTable.get_by_atomic_number(CARBON)
+	const NITROGEN: int = 7
+	var nitrogen_data: ElementData = PeriodicTable.get_by_atomic_number(NITROGEN)
+	const OXIGEN: int = 8
+	var oxygen_data: ElementData = PeriodicTable.get_by_atomic_number(OXIGEN)
+	var workspace_context: WorkspaceContext = MolecularEditorContext.get_current_workspace_context() as WorkspaceContext
+	var context: StructureContext = null
+	var atomic_structure: AtomicStructure = null
+	var candidate_element: ElementData
+	for candidate: AtomCandidate in _candidates:
+		if context == null or context.nano_structure.int_guid != candidate.structrure_id:
+			# OPTIMIZATION: Candidates are meant to be grouped by structure,
+			# so this if condition should not be needed to run often
+			context = workspace_context.get_structure_context(candidate.structrure_id)
+			atomic_structure = context.nano_structure as AtomicStructure
+		var pos_2d: Vector2 = _camera.unproject_position(candidate.atom_position)
```

Comment:
You can skip the whole iteration if the candidate is not visible, by adding this condition right before that line:

```gdscript
if not _camera.is_position_in_frustrum(camdidate.atom_position):
    continue
```

The only downside is if you have a selected atom near the edge of the screen and the candidate is outside, it won't draw the connection to that candidate and it might mislead the user into thinking that candidate doesn't exists, but on the other hand, it saves a ton of draw calls if the selection is large but not entirely visible

Reply:
turns out is **frustum**, not frustrum. My entire life is a lie...

Reply:
Oops, I made a few typo in that comment :sweat_smile: